### ISSUE TITLE
panda dies when asked to install a module that is already installed

### ISSUE THREAD

**Issue by autarch** [tadzik/panda#269]
If I run `panda install URI` twice I get this:

```
==> Installing URI
URI:ver<v0.1.1>:auth<>:api<> already installed
  in block  at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/FECC4E070CD72F99217622228852CF7CA2014157 line 61
  in sub indir at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/282870C07C23A46BBDC1336ACA09900A0F730A1D line 20
  in method install at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/FECC4E070CD72F99217622228852CF7CA2014157 line 42
  in method install at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/B4AF50D01531A0370B8163016E0770BF0EF315ED line 146
  in method resolve at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/B4AF50D01531A0370B8163016E0770BF0EF315ED line 219
  in sub MAIN at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/resources/794E4B4C32A90D283D02A11D5C3CA4A31593B0E9 line 18
  in block <unit> at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/resources/794E4B4C32A90D283D02A11D5C3CA4A31593B0E9 line 150

Actually thrown at:
  in method install at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/FECC4E070CD72F99217622228852CF7CA2014157 line 44
  in method install at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/B4AF50D01531A0370B8163016E0770BF0EF315ED line 146
  in method resolve at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/sources/B4AF50D01531A0370B8163016E0770BF0EF315ED line 219
  in sub MAIN at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/resources/794E4B4C32A90D283D02A11D5C3CA4A31593B0E9 line 18
  in block <unit> at /home/autarch/.rakudobrew/moar-nom/install/share/perl6/site/resources/794E4B4C32A90D283D02A11D5C3CA4A31593B0E9 line 150
```

**Comment by niner** [comment]
That's expected. Dists are supposed to be immutable. Once you give it a version, it should never change anymore, so what's the point of installing, if exactly the same is already there?

In case you actually know that you still need to install again (like when you're developing said module or try to debug an installations issue) you can override the check using panda --force install URI

**Comment by niner** [comment]
Reopened, because the error message can clearly be improved and should mention most of what I answered here.

**Comment by autarch** [comment]
While of course we don't want to reinstall without a `--force` flag, having panda die is not ok, as that aborts installing other modules give on the CLI (e.g. `panda install URI File::Temp`). If URI is already installed in that case, panda should still install File::Temp.

**Comment by FROGGS** [comment]
Yes, it should mention that `X` is already installed in `panda install X`. And it should offer a reinstall by mentioning `--force`.

**Comment by tadzik** [comment]
Right now panda's not very useful for module development/debugging purposes, where you may want to repeatedly install a module while not necessarily bumping the version number every single time. Does CUR give us a way to actually force the module in?

**Comment by lizmat** [comment]
> On 25 Dec 2015, at 21:08, Tadeusz Sośnierz <notifications@github.com> wrote:
> 
> Right now panda's not very useful for module development/debugging purposes, where you may want to repeatedly install a module while not necessarily bumping the version number every single time. Does CUR give us a way to actually force the module in?

If you compile a new version of rakudo, all previously installed modules ARE STILL INSTALLED!

The first time you try to access them, they will be pre-compiled again automatically.



Liz

**Comment by tadzik** [comment]
I'm aware of that; it's problematic not when reinstalling rakudo, it's 
when you make subtle changes in a module and want to install it globally 
for further testing/debugging without bumping the version number. This 
is not a hypothetical scenario, I was going through this multiple times 
just yesterday, partially due to how fragile the whole system currently is.

**Comment by lizmat** [comment]
Ok, gotcha now.

Pretty sure we won’t resolve this before 6.c / 2015.12 is out.  Should be possible to do soon afterwards, I would think.


Liz

**Comment by niner** [comment]
Err... as I said in my second comment: panda --force install URI will happily ignore the "already installed" check and that's exactly what I use when testing/debugging installation issues.

**Comment by hankache** [comment]
@niner --force is not working anymore. i just nuked everything and built from scratch. I can't install File::Find or JSON::Fast or anything that panda depends on. Using --force gave the same error:
`JSON::Fast:ver<v0.3>:auth<>:api<> already installed`

**Comment by colomon** [comment]
Just to double check, you did

    panda --force install JSON::Fast

rather than

    panda install --force JSON::Fast

right?  I know I've done the later quite a few times, and I'm pretty sure
it just silently ignores the --force if it comes after the install.

--Sol

On Thu, Dec 31, 2015 at 9:29 AM, Naoum Hankache <notifications@github.com>
wrote:

> @niner <https://github.com/niner> --force is not working anymore. i just
> nuked everything and built from scratch. I can't install File::Find or
> JSON::Fast or anything that panda depends on. Using --force gave the same
> error:
> JSON::Fast:ver<v0.3>:auth<>:api<> already installed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tadzik/panda/issues/269#issuecomment-168204091>.
>



-- 
Solomon Foster: colomon@gmail.com
HarmonyWare, Inc: http://www.harmonyware.com

**Comment by hankache** [comment]
@colomon yes i did `panda --force install JSON::Fast`

**Comment by zoffixznet** [comment]
To add: it shouldn't bail out from the install, it should just say just "Already latest version" and proceed with installing the next module (or dependency) it was asked to install.

**Comment by shadowcat-mst** [comment]
Right, it manages it fine when it's a dependency, it skips already present ones. Just need to add the same logic to the top level, probably by treating the list to 'panda install' as a user supplied dep set

**Comment by shadowcat-mst** [comment]
As of rakudo/rakudo@2ae0dfa --force should work again while we figure out how to fix this properly.

**Comment by niner** [comment]
I'm working on a Perl 6 API that can be used by panda to ask if a module is already installed: https://github.com/rakudo/rakudo/commit/92ef589c74b6a4a054d9672c5de907e43d3765da

I see panda's state file where it records, which modules it installed as a prime source of this issue, since it's redundant information and therefore prone to get out of sync. Replacing it by actually querying the prime source of information seems to be the way to go.

Please comment on this plan and especially the proposed API
